lib/smack/SmackRep.cpp

```diff
@@ -659,9 +659,55 @@ const Expr* SmackRep::lit(const llvm::Value* v, bool isUnsigned) {
       }
       const APInt API = APF.bitcastToAPInt();
       const APInt n_sign = API.trunc(expSize+sigSize-1);
-      const APInt sig = n_sign.trunc(sigSize-1);
-      const APInt exp = n_sign.lshr(sigSize-1);
-      return Expr::lit(APF.isNegative(), sig.toString(10, false), exp.toString(10, false), sigSize, expSize);
+      const APInt exp = n_sign.lshr(sigSize-1).trunc(expSize);
+      APInt sig = n_sign.trunc(sigSize-1).zext(sigSize);
+
+      if (exp.isAllOnesValue()) {
+        std::string v;
+        if (sig == 0) {
+          if (API.isNegative()) {
+            v = "-oo";
+          } else {
+            v = "+oo";
+          }
+        } else {
+          v = "NaN";
+        }
+        return Expr::lit(v, sigSize, expSize);
+      }
+
+      if (exp != 0) {
+        sig.setBit(sigSize - 1); //hidden bit
+      }
+
+      APInt bias = APInt::getSignedMaxValue(expSize);
+      if (exp == 0) {
+        --bias;
+      }
+
+      bool ov;
+
+      APInt moveDec = exp.usub_ov(bias, ov);
```

Comment:
Do you need to check for overflow here? Can it ever happen? Would it be really bad if it happens?
I guess I would consider adding `assert(!overflow && "Overflow in floating-point computation");` here.
And at the other places where `overflow` gets used.

Reply:
The only other way to subtract two APInts is to use the -= operator. This would require me to write:
```
APInt moveDec = exp;
moveDec -= bias;
```
I wanted to keep it all on one line, so I used `usub_ov`. There's another function called usub_sat that doesn't take the overflow argument, but for some reason it's not visible to SMACK.

Reply:
To answer your question, overflow happens whenever `exp < bias`, which is quite often, but it doesn't matter at all, because we want the value to wrap around anyways.